python/ray/data/read_api.py

```diff
@@ -2583,6 +2584,22 @@ def read_hudi(
     )
 
 
+@PublicAPI
+def from_daft(df: "daft.DataFrame") -> MaterializedDataset:
+    """Create a :class:`~ray.data.Dataset` from a `Daft DataFrame <https://www.getdaft.io/projects/docs/en/stable/api_docs/dataframe.html>`_.
+
+    Args:
+        df: A Daft DataFrame
+
+    Returns:
+        A :class:`~ray.data.MaterializedDataset` holding rows read from the DataFrame.
+    """
+    # NOTE: Today this returns a MaterializedDataset. We should also integrate Daft such that we can stream object references into a Ray
+    # dataset. Unfortunately this is very tricky today because of the way Ray Datasources are implemented with a fully-materialized `list`
+    # of ReadTasks, rather than an iterator which can lazily return these tasks.
```

Comment:
> the way Ray Datasources are implemented with a fully-materialized `list`
     # of ReadTasks, rather than an iterator which can lazily return these tasks.

Can you elaborate  on why it's hard to determine the # of read tasks? 
Btw, we do plan to support an iterator, in order to support unbounded data sources. 
But IIUC, a daft DataFrame should be bounded?

Reply:
makes sense. thanks. 
Nit, can you update the return type to just `Dataset`. `MaterializedDataset` is an implementation detail that may change in the future.

Reply:
@jaychia `MaterializedDataset` is meant to store fully _materialized_ data hence the limiting form-factor of the current abstraction.

We will be providing an iterator based operator sourcing data from predefined operator. Please feel free to file a ticket or take a stab at it.